### ISSUE TITLE
Consider Updating Festival

### ISSUE THREAD

**Issue by greg-hellings** [crosswire/xiphos#1023]
The version of Festival that we package is currently 1.9.6.

The last release (2014) was 2.4.0. We should consider updating the bundled package and testing that it still works OK.

**Comment by greg-hellings** [comment]
Looks like I'm wrong. 2.5 was released in 2017. http://festvox.org/festival/

**Comment by alerque** [comment]
Is there any possibility of unbundling it?

**Comment by greg-hellings** [comment]
I doubt it, but that's part of why I put this here. It depends on how the package looks and works. Fedora doesn't like to package end-user MinGW executables. Only libraries and - with a few very minor exceptions - some utilities here and there. One example is the gdb.exe that we package. I also was able to slip in the SWORD utils, because they are just part of the package and aren't the main target of the project.

If Festival can reasonably be passed off as a shared library, then it's possible. But if not, then we'll have to continue bundling it in.

**Comment by bgermann** [comment]
The xiphos Debian package does not include the bundled festival but recommends the festival Debian package. I tested it with version 2.5.0, so it should be fine to update or unbundle.

**Comment by LAfricain** [comment]
It seems possible to make progress on this issue and solve #762. If pico2wave (svox-pico) really was a bad idea, there are an other strategy. Instead of updating Festival, we could adopt the following strategy:

### Use the operating system’s native APIs (highly recommended)
Instead of forcing Xiphos to use festival we can delegate this task to the operating system’s built-in speech capabilities. This is how modern applications generally handle text-to-speech.
*   **On Linux:** Use **Speech Dispatcher** (`libspeechd`). Xiphos simply sends the text to the dispatcher, and the operating system handles the reading using the engine the user has natively installed.
*   **On Windows:** Call the native **SAPI 5** or **WinRT TTS** APIs. Windows already includes high-quality built-in voices, requiring no additional dependencies to compile or include in the Xiphos installer.
*   **On macOS:** Call `NSSpeechSynthesizer`.
*   *Benefits:* No extra bloat, no dependency headaches, and you get high-quality, out-of-the-box system voices.

**Comment by LAfricain** [comment]
But if we want to go with this solution, we should document it properly in an Help section. Personally, I think it's great. Here's a draft of the documentation; I included Mac even though Xiphos isn't available on Mac yet.

### Proposed documentation for the Xiphos Help Manual:

# Text-to-Speech (TTS) / Read Aloud Feature

To provide the best possible experience and keep the application lightweight,  **Xiphos relies on the native speech technologies built into your operating system**. 

This approach allows you to use the high-quality voices already installed on your computer and gives you full control over which voice to use.

Here is how to configure the voice depending on your operating system:

### 🪟 Windows Users
On Windows, the TTS feature works right out of the box. Xiphos uses the native Windows Speech API (SAPI), allowing you to enjoy the high-quality voices provided by Microsoft.

*   **To change the voice or reading speed:**
    1. Open the Windows **Start** menu and go to **Settings**.
    2. Navigate to **Time & language** > **Speech**.
    3. Scroll down to the "Voices" section. Here, you can select a different voice (e.g., change from a male to a female voice) and adjust the speaking speed. Xiphos will automatically use your new system default.

### 🐧 Linux Users (Ubuntu, Fedora, etc.)
On Linux, Xiphos delegates speech management to the system standard: **Speech Dispatcher** (`libspeechd`). This means your Linux system is the one deciding which vocal engine to use, not Xiphos.

Depending on how your system is configured, you have several engine options available in your distribution's repositories:

*   **eSpeak / eSpeak-NG (Default):** Often the default engine on many Linux distributions. It is extremely lightweight, fast, and supports many languages, but the voice sounds highly robotic.
*   **Festival:** A classic, highly customizable open-source engine developed by the University of Edinburgh. You can install it via your package manager (e.g., `sudo apt install festival`).
*   **SVOX Pico (More natural voice):** A lightweight engine that sounds much more natural than eSpeak. To use it:
    1. On Ubuntu/Debian, open a terminal and install the package: 
       `sudo apt install libttspico-utils`
    2. Run the Speech Dispatcher configuration tool by typing:
       `spd-conf`
    3. Follow the on-screen instructions. When asked for the *Default output module*, type `pico`. Xiphos will now use this smoother voice.
*   **Piper (High-quality Neural voices - Advanced):** If you want ultra-realistic, offline AI voices, you can configure your Linux system to use **Piper**. Since Piper is not in the official `apt` repositories, you will need to manually download the Piper binary and configure a "generic" module in your Speech Dispatcher's `speechd.conf` file.

### 🍏 macOS Users
On macOS, Xiphos uses the excellent, natural-sounding voices provided natively by Apple.

*   **To change the voice:**
    Go to **System Settings** > **Accessibility** > **Spoken Content**. There, you can change the default system voice, adjust the speaking rate, and download new premium voices directly from Apple's servers.

**Comment by alerque** [comment]
@LAfricain Are you copy/pasting LLM output here?

**Comment by LAfricain** [comment]
@alerque yes, I often ask the AI to summarize what I found directly in English.

**Comment by LAfricain** [comment]
I made some changes (my local git) so we can use the text-to-speech engine of our choice. So now I can use a French-language engine. That's really great! However, while the text-to-speech option in the View menu works very well, the right-click option is faulty. It only reads the selection on the second selection. Before that, it reads what’s in the dictionary field. This is already strange behavior; my changes aren’t the cause, because it does the same thing with Festival in the current version of Xiphos.

There’s quite a bit of work to be done in the TTS area; we could add a button or a shortcut to pause and resume playback. We could also have it read from a verse without stopping, rather than reading just a single verse...

**Comment by karlkleinpaste** [comment]
It is probably the same problem as the "click word to dictionary" problem solved a while back, using the wrong source for the text.

Read Aloud is literally the very first feature I added to then-GnomeSword in 2006. I was woefully ignorant of much of the code, unaware of many details of the full Sword interface, and barely cognizant of festival itself. It was surprising to me beyond measure when the feature was well-received, considering that I did it as a first-time, get-your-feet-wet, experimental hack. That it has needed to be updated has been clear for years.

**Comment by LAfricain** [comment]
Okay, so do you agree with the proposal to hand control of the TTS back to the system? If so, I already have a branch with the solution, but we’ll need to document it in the help section and test it on Windows. However, we’ll need to take it to the next level to enable support for new languages. After that, we can add other features. 
Since we're on the subject, I'd like to ask you how you'd prefer me to proceed: should I push my branch to master, or send you the entire branch?

**Comment by LAfricain** [comment]
I just noticed that right-clicking on selected text to read it aloud causes Xiphos to crash.

**Comment by karlkleinpaste** [comment]
Not so, in my fedora 42 build. I can sweep text, right-click, select Read Aloud, and festival speaks. Not a problem. But it is using the previous selection instead of the current one, which is I surmise as the same error as "click word to dictionary" recently fixed.

**Comment by LAfricain** [comment]
I'm sorry I was speaking about windows, not Ubuntu!
